### PR TITLE
random miscellaneous lint fixes

### PR DESCRIPTION
**Pull request by eli-schwartz**
- add a bunch of mypy annotations
- take some imports that aren't used except for typing, and move them to TYPE_CHECKING
- fix some broken or redundant logic uncovered by mypy:
  - `if func:`
  - checking `if isinstance` that can never be false
- use another couple of dataclasses

### CODE REVIEW

File: mesonbuild/backend/ninjabackend.py [hunk 1]
```diff
@@ -3605,7 +3607,7 @@ def generate_gcov_clean(self) -> None:
         gcda_elem.add_item('description', 'Deleting gcda files')
         self.add_build(gcda_elem)
 
-    def get_user_option_args(self):
+    def get_user_option_args(self) ->  T.List[str]:
```
Comment (cclauss) [inline mesonbuild/backend/ninjabackend.py:3610]:
```suggestion
    def get_user_option_args(self) -> T.List[str]:
```

File: mesonbuild/build.py [hunk 2]
```diff
@@ -721,6 +721,9 @@ class BuildTarget(Target):
         ('cpp', 'cuda'),
     ])
 
+    # This is used by the backend to cache complex computation.
+    cached_generated_headers: T.Optional[T.List[FileOrString]]
```
Comment (dcbaker) [inline mesonbuild/build.py:725]: I have some patches somewhere to stop adding this to the Target and just store it in the NijnjaBackend. I'm fine with this being here for the short term, but it's a really ugly hack and we should not do it in the long run.

File: mesonbuild/compilers/compilers.py [hunk 3]
```diff
@@ -406,34 +406,27 @@ def get_base_link_args(options: 'KeyedOptionDictType', linker: 'Compiler',
...
-        self.stdout = stdo
-        self.stderr = stde
-        self.input_name = input_name
-        self.output_name = output_name
-        self.command = command or []
-        self.cached = cached
-        self.returncode = returncode
+    stdout: T.Optional[str] = None
+    stderr: T.Optional[str] = None
+    command: T.Optional[T.List[str]] = field(default_factory=list)
```
Comment (dcbaker) [inline mesonbuild/compilers/compilers.py]: do we really want this to be optional now that it's default initialized to []?

**Review by dcbaker — APPROVED** [review]
There's one issue I spotted that isn't super important to since it gets fixed in the next patch. Otherwise LGTM

File: mesonbuild/compilers/compilers.py (hunk 3 above)
Comment (eli-schwartz) [reply mesonbuild/compilers/compilers.py]: You're right, the optionality should really be removed in the first commit.

File: mesonbuild/build.py (hunk 2 above)
Comment (eli-schwartz) [reply mesonbuild/build.py:725]: Yup, it's incredibly ugly as can also be seen by the fact that hasattr is ever used at all. :D

**Pull request merged**